Library/Homebrew/tap.rb

```diff
@@ -36,28 +39,30 @@ class Tap
     #{HOMEBREW_TAP_PYPI_FORMULA_MAPPINGS}
   ].freeze
 
-  def self.fetch(*args)
-    case args.length
-    when 1
-      user, repo = args.first.split("/", 2)
-    when 2
-      user = args.first
-      repo = args.second
-    end
+  # sig { params(user: String, repo: String).returns(T.attached_class) }
+  def self.fetch(user, repo = T.unsafe(nil))
+    user, repo = user.split("/", 2) if repo.nil?
 
-    raise "Invalid tap name '#{args.join("/")}'" if [user, repo].any? { |part| part.nil? || part.include?("/") }
+    if [user, repo].any? { |part| part.nil? || part.include?("/") }
+      raise ArgumentError, "Invalid tap name: '#{[*user, *repo].join("/")}'"
+    end
 
     # We special case homebrew and linuxbrew so that users don't have to shift in a terminal.
-    user = user.capitalize if ["homebrew", "linuxbrew"].include? user
+    user = user.capitalize if ["homebrew", "linuxbrew"].include?(user)
     repo = repo.sub(HOMEBREW_OFFICIAL_REPO_PREFIXES_REGEX, "")
 
     return CoreTap.instance if ["Homebrew", "Linuxbrew"].include?(user) && ["core", "homebrew"].include?(repo)
     return CoreCaskTap.instance if user == "Homebrew" && repo == "cask"
 
-    cache_key = "#{user}/#{repo}".downcase
-    cache.fetch(cache_key) { |key| cache[key] = Tap.new(user, repo) }
+    _fetch(user, repo)
+  end
+
+  sig { params(user: String, repo: String).returns(T.attached_class) }
+  private_class_method cached_class_method def self._fetch(user, repo)
+    new(user, repo)
   end
```

Comment:
If possible, I think we should avoid making this a pattern. Ideally the original method would be able to be cached directly.